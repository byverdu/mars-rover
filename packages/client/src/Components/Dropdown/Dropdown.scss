$purple: #5b13ec;
$white: #ffffff;

.wrapper {
  height: 60px;
  border: 1px solid grey;
  position: relative;

  &:before {
    content: '\2026';
    position: absolute;
    transform: rotate(90deg);
    left: 20px;
    font-size: 40px;
    top: 5px;
  }

  h5 {
    margin: 0;
    padding: 23px 0 20px;
  }

  .items {
    justify-content: center;
    position: relative;
    align-items: center;
    padding: 20px;
    border: 1px solid grey;
    background: white;
    border-radius: 0 0 10px 10px;
    top: 1px;
    width: 96%;
    margin: 0 auto;
    z-index: 2;

    &.hide {
      display: none;
    }

    &.show {
      display: flex;
    }

    .coords {
      display: flex;
      flex-wrap: wrap;
      background-color: $white;
      color: $purple;
      z-index: 1;
    }

    .item-coords {
      width: 100px;
      box-shadow: 1px -1px, -1px 1px;
      position: relative;
      display: flex;
      justify-content: center;
      align-items: center;
    }

    .cardinals {
      display: flex;
      flex-direction: column;
      background-color: $white;
      color: $purple;
      margin-left: 10px;
    }

    .item-cardinals {
      box-shadow: 1px -1px, -1px 1px;
      padding: 10px;
    }

    .selected {
      background-color: $purple;
      color: $white;
    }
  }
}
