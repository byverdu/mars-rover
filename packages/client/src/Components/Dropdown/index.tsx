import React, { useState } from 'react';
import './Dropdown.scss';
import { EnumCardinalPoints } from '../../Models/enums';

const classnames = require('classnames');

export interface DropdownProps {
  width: number;
  height: number;
  setRoverPosition: (value: string) => void;
}

function selectedElement(selector): HTMLElement {
  return Array.from(document.querySelectorAll(selector)).find(
    (item: HTMLElement) => item.className.includes('selected')
  );
}

function getTextSiblingSelected(selector): string {
  const selectedElement = document.querySelector(selector);
  return selectedElement ? selectedElement.textContent : '';
}

function coordsBuilder(width, height, setRoverPosition) {
  const wrapper = [];
  function spanBuilder(newHeight) {
    for (let i = 0; i < width; i += 1) {
      wrapper.push(
        <span
          key={`${i}-${height - newHeight}`}
          className="item-coords"
          onClick={(e: React.MouseEvent<HTMLElement>) => {
            e.preventDefault();
            const alreadySelected = selectedElement('.coords .item-coords');

            if (alreadySelected) {
              alreadySelected.classList.remove('selected');
            }

            const element = e.target as HTMLElement;
            const siblingText = getTextSiblingSelected(
              '.item-cardinals.selected'
            );
            element.classList.add('selected');

            setRoverPosition(`${element.textContent} ${siblingText}`);
          }}
        >
          {i} {height - newHeight}
        </span>
      );
    }
  }
  let counter = 1;
  while (counter <= height) {
    spanBuilder(counter);
    counter += 1;
  }

  return wrapper;
}

function cardinalPointsBuilder(setRoverPosition) {
  return [
    EnumCardinalPoints.N,
    EnumCardinalPoints.S,
    EnumCardinalPoints.E,
    EnumCardinalPoints.W
  ].map((cardinal) => (
    <span
      key={cardinal}
      className="item-cardinals"
      onClick={(e: React.MouseEvent<HTMLElement>) => {
        e.preventDefault();
        const alreadySelected = selectedElement('.cardinals .item-cardinals');
        if (alreadySelected) {
          alreadySelected.classList.remove('selected');
        }
        const element = e.target as HTMLElement;
        const siblingText = getTextSiblingSelected('.item-coords.selected');

        element.classList.add('selected');
        setRoverPosition(`${siblingText} ${element.textContent}`);
      }}
    >
      {cardinal}
    </span>
  ));
}

const CoordsDropdown: React.FC<DropdownProps> = ({
  width,
  height,
  setRoverPosition
}) => {
  const [visibility, setVisibility] = useState(true);
  const itemsClassnames = classnames('items', {
    show: visibility,
    hide: !visibility
  });
  const onWrapperClick = () => setVisibility(!visibility);

  return (
    <div className="wrapper">
      <h5 onClick={onWrapperClick}>Set Rover Coords</h5>
      <div className={itemsClassnames}>
        <div className="coords" style={{ width: `${width * 100}px` }}>
          {coordsBuilder(width, height, setRoverPosition)}
        </div>
        <div className="cardinals">
          {cardinalPointsBuilder(setRoverPosition)}
        </div>
      </div>
    </div>
  );
};

export default CoordsDropdown;
