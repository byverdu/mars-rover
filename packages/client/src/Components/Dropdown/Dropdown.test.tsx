import React from 'react';
import Dropdown, { DropdownProps } from './index';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

const props: DropdownProps = {
  width: 8,
  height: 8,
  setRoverPosition: () => {}
};

describe('<Dropdown />', () => {
  const wrapper = mount(<Dropdown {...props} />);

  it('should create a snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
