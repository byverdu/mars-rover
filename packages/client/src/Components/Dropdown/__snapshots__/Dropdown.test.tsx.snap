// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should create a snapshot 1`] = `
<CoordsDropdown
  height={8}
  setRoverPosition={[Function]}
  width={8}
>
  <div
    className="wrapper"
  >
    <h5
      onClick={[Function]}
    >
      Set Rover Coords
    </h5>
    <div
      className="items show"
    >
      <div
        className="coords"
        style={
          Object {
            "width": "800px",
          }
        }
      >
        <span
          className="item-coords"
          key="0-7"
          onClick={[Function]}
        >
          0
           
          7
        </span>
        <span
          className="item-coords"
          key="1-7"
          onClick={[Function]}
        >
          1
           
          7
        </span>
        <span
          className="item-coords"
          key="2-7"
          onClick={[Function]}
        >
          2
           
          7
        </span>
        <span
          className="item-coords"
          key="3-7"
          onClick={[Function]}
        >
          3
           
          7
        </span>
        <span
          className="item-coords"
          key="4-7"
          onClick={[Function]}
        >
          4
           
          7
        </span>
        <span
          className="item-coords"
          key="5-7"
          onClick={[Function]}
        >
          5
           
          7
        </span>
        <span
          className="item-coords"
          key="6-7"
          onClick={[Function]}
        >
          6
           
          7
        </span>
        <span
          className="item-coords"
          key="7-7"
          onClick={[Function]}
        >
          7
           
          7
        </span>
        <span
          className="item-coords"
          key="0-6"
          onClick={[Function]}
        >
          0
           
          6
        </span>
        <span
          className="item-coords"
          key="1-6"
          onClick={[Function]}
        >
          1
           
          6
        </span>
        <span
          className="item-coords"
          key="2-6"
          onClick={[Function]}
        >
          2
           
          6
        </span>
        <span
          className="item-coords"
          key="3-6"
          onClick={[Function]}
        >
          3
           
          6
        </span>
        <span
          className="item-coords"
          key="4-6"
          onClick={[Function]}
        >
          4
           
          6
        </span>
        <span
          className="item-coords"
          key="5-6"
          onClick={[Function]}
        >
          5
           
          6
        </span>
        <span
          className="item-coords"
          key="6-6"
          onClick={[Function]}
        >
          6
           
          6
        </span>
        <span
          className="item-coords"
          key="7-6"
          onClick={[Function]}
        >
          7
           
          6
        </span>
        <span
          className="item-coords"
          key="0-5"
          onClick={[Function]}
        >
          0
           
          5
        </span>
        <span
          className="item-coords"
          key="1-5"
          onClick={[Function]}
        >
          1
           
          5
        </span>
        <span
          className="item-coords"
          key="2-5"
          onClick={[Function]}
        >
          2
           
          5
        </span>
        <span
          className="item-coords"
          key="3-5"
          onClick={[Function]}
        >
          3
           
          5
        </span>
        <span
          className="item-coords"
          key="4-5"
          onClick={[Function]}
        >
          4
           
          5
        </span>
        <span
          className="item-coords"
          key="5-5"
          onClick={[Function]}
        >
          5
           
          5
        </span>
        <span
          className="item-coords"
          key="6-5"
          onClick={[Function]}
        >
          6
           
          5
        </span>
        <span
          className="item-coords"
          key="7-5"
          onClick={[Function]}
        >
          7
           
          5
        </span>
        <span
          className="item-coords"
          key="0-4"
          onClick={[Function]}
        >
          0
           
          4
        </span>
        <span
          className="item-coords"
          key="1-4"
          onClick={[Function]}
        >
          1
           
          4
        </span>
        <span
          className="item-coords"
          key="2-4"
          onClick={[Function]}
        >
          2
           
          4
        </span>
        <span
          className="item-coords"
          key="3-4"
          onClick={[Function]}
        >
          3
           
          4
        </span>
        <span
          className="item-coords"
          key="4-4"
          onClick={[Function]}
        >
          4
           
          4
        </span>
        <span
          className="item-coords"
          key="5-4"
          onClick={[Function]}
        >
          5
           
          4
        </span>
        <span
          className="item-coords"
          key="6-4"
          onClick={[Function]}
        >
          6
           
          4
        </span>
        <span
          className="item-coords"
          key="7-4"
          onClick={[Function]}
        >
          7
           
          4
        </span>
        <span
          className="item-coords"
          key="0-3"
          onClick={[Function]}
        >
          0
           
          3
        </span>
        <span
          className="item-coords"
          key="1-3"
          onClick={[Function]}
        >
          1
           
          3
        </span>
        <span
          className="item-coords"
          key="2-3"
          onClick={[Function]}
        >
          2
           
          3
        </span>
        <span
          className="item-coords"
          key="3-3"
          onClick={[Function]}
        >
          3
           
          3
        </span>
        <span
          className="item-coords"
          key="4-3"
          onClick={[Function]}
        >
          4
           
          3
        </span>
        <span
          className="item-coords"
          key="5-3"
          onClick={[Function]}
        >
          5
           
          3
        </span>
        <span
          className="item-coords"
          key="6-3"
          onClick={[Function]}
        >
          6
           
          3
        </span>
        <span
          className="item-coords"
          key="7-3"
          onClick={[Function]}
        >
          7
           
          3
        </span>
        <span
          className="item-coords"
          key="0-2"
          onClick={[Function]}
        >
          0
           
          2
        </span>
        <span
          className="item-coords"
          key="1-2"
          onClick={[Function]}
        >
          1
           
          2
        </span>
        <span
          className="item-coords"
          key="2-2"
          onClick={[Function]}
        >
          2
           
          2
        </span>
        <span
          className="item-coords"
          key="3-2"
          onClick={[Function]}
        >
          3
           
          2
        </span>
        <span
          className="item-coords"
          key="4-2"
          onClick={[Function]}
        >
          4
           
          2
        </span>
        <span
          className="item-coords"
          key="5-2"
          onClick={[Function]}
        >
          5
           
          2
        </span>
        <span
          className="item-coords"
          key="6-2"
          onClick={[Function]}
        >
          6
           
          2
        </span>
        <span
          className="item-coords"
          key="7-2"
          onClick={[Function]}
        >
          7
           
          2
        </span>
        <span
          className="item-coords"
          key="0-1"
          onClick={[Function]}
        >
          0
           
          1
        </span>
        <span
          className="item-coords"
          key="1-1"
          onClick={[Function]}
        >
          1
           
          1
        </span>
        <span
          className="item-coords"
          key="2-1"
          onClick={[Function]}
        >
          2
           
          1
        </span>
        <span
          className="item-coords"
          key="3-1"
          onClick={[Function]}
        >
          3
           
          1
        </span>
        <span
          className="item-coords"
          key="4-1"
          onClick={[Function]}
        >
          4
           
          1
        </span>
        <span
          className="item-coords"
          key="5-1"
          onClick={[Function]}
        >
          5
           
          1
        </span>
        <span
          className="item-coords"
          key="6-1"
          onClick={[Function]}
        >
          6
           
          1
        </span>
        <span
          className="item-coords"
          key="7-1"
          onClick={[Function]}
        >
          7
           
          1
        </span>
        <span
          className="item-coords"
          key="0-0"
          onClick={[Function]}
        >
          0
           
          0
        </span>
        <span
          className="item-coords"
          key="1-0"
          onClick={[Function]}
        >
          1
           
          0
        </span>
        <span
          className="item-coords"
          key="2-0"
          onClick={[Function]}
        >
          2
           
          0
        </span>
        <span
          className="item-coords"
          key="3-0"
          onClick={[Function]}
        >
          3
           
          0
        </span>
        <span
          className="item-coords"
          key="4-0"
          onClick={[Function]}
        >
          4
           
          0
        </span>
        <span
          className="item-coords"
          key="5-0"
          onClick={[Function]}
        >
          5
           
          0
        </span>
        <span
          className="item-coords"
          key="6-0"
          onClick={[Function]}
        >
          6
           
          0
        </span>
        <span
          className="item-coords"
          key="7-0"
          onClick={[Function]}
        >
          7
           
          0
        </span>
      </div>
      <div
        className="cardinals"
      >
        <span
          className="item-cardinals"
          key="N"
          onClick={[Function]}
        >
          N
        </span>
        <span
          className="item-cardinals"
          key="S"
          onClick={[Function]}
        >
          S
        </span>
        <span
          className="item-cardinals"
          key="E"
          onClick={[Function]}
        >
          E
        </span>
        <span
          className="item-cardinals"
          key="W"
          onClick={[Function]}
        >
          W
        </span>
      </div>
    </div>
  </div>
</CoordsDropdown>
`;
